### PR TITLE
Shield rework

### PR DESCRIPTION
**Pull request by Thrifinesma**

### CODE REVIEW

File: src/Misc/Shield.cpp [hunk 1]
```diff
@@ -12,93 +12,90 @@
...
+    this->DrawShield();
 }
 
-const TechnoTypeExt::ExtData* ShieldTechnoClass::GetExt()
+const TechnoTypeExt::ExtData* ShieldTechnoClass::GetTechnoTypeExt()
 {
     return TechnoTypeExt::ExtMap.Find(this->Techno->GetTechnoType());
 }
 
+#define Serialize(Stm) \
```
Comment (Metadorius) [inline src/Misc/Shield.cpp:39]: Why a define instead of a proper function like what's used in Ext classes?

File: src/Enum/ShieldType.cpp [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+{
+	return "ShieldTypes";
+}
+
+void ShieldTypeClass::AddDefaults()
+{
+	FindOrAllocate(NONE_STR);
+}
+
+void ShieldTypeClass::LoadListSection(CCINIClass* pINI)
```
Comment (Metadorius) [inline src/Enum/ShieldType.cpp:15]: This should be implemented in `Enumerable<T>`. Actually I have already fixed the implementation in laser trails branch so I guess I could cherrypick it to develop and merge the changes here.

File: src/Enum/ShieldType.cpp [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+	for (int i = 0; i < pINI->GetKeyCount(GetMainSection()); ++i)
+	{
+		if (pINI->ReadString(GetMainSection(), pINI->GetKeyName(GetMainSection(), i), "", Phobos::readBuffer))
+		{
+			FindOrAllocate(Phobos::readBuffer);
+			Debug::Log("ShieldTypeClass :: LoadListSection check [%s] \n", Phobos::readBuffer);
+		}
+	}
+
+	for (auto& radType : Array)
```
Comment (Metadorius) [inline src/Enum/ShieldType.cpp:26]: CnP naming error.

File: src/Ext/Techno/Hooks.Shield.cpp [hunk 4]
```diff
@@ -138,22 +138,38 @@ DEFINE_HOOK(6F36DB, TechnoClass_WhatWeaponShouldIUse_Shield, 8)
...
-	if (pExt->ShieldData)
-		pExt->ShieldData->AI();
+	if (const auto pShieldData = pExt->ShieldData.get())
+		pShieldData->AI();
 
 	return 0;
 }
 
+// Ares-hook jmp to this offset
+DEFINE_HOOK(71A88D, TemporalClass_AI_Shield, 0)
```
Comment (Metadorius) [inline src/Ext/Techno/Hooks.Shield.cpp:154]: Missing newlines and it's not clear what's happening without enum'ed addresses and/or comments.

File: src/Ext/TechnoType/Body.cpp [hunk 5]
```diff
@@ -15,8 +15,7 @@ TechnoTypeExt::ExtContainer TechnoTypeExt::ExtMap;
 
 void TechnoTypeExt::ExtData::Initialize()
 {
-	auto pThis = this->OwnerObject();
-	UNREFERENCED_PARAMETER(pThis);
+	this->Shield = ShieldTypeClass::FindOrAllocate("Radiation");
```
Comment (Metadorius) [inline src/Ext/TechnoType/Body.cpp:18]: CnP naming error yet again. Not sure if we even need the default stuff. See lasertrails branch.

File: src/ExtraHeaders/ArmorType.h [hunk 6]
```diff
@@ -0,0 +1,19 @@
+#pragma once
+
+#include <Utilities/Template.h>
+
+
+class ArmorType : public Valueable<int>
+{
+public:
+	template <typename T>
+	ArmorType(T value) {
```
Comment (Metadorius) [inline src/ExtraHeaders/ArmorType.h:10]: Brace not on a newline.

File: src/ExtraHeaders/ArmorType.h [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+		this->Value = (int)std::move(value);
+	}
+
+	template <typename T>
+	ArmorType& operator = (T value) {
```
Comment (Metadorius) [inline src/ExtraHeaders/ArmorType.h:15]: Ditto.

File: src/Misc/Shield.cpp [hunk 8]
```diff
@@ -11,77 +11,76 @@
...
+}
 
-	pToExt->ShieldData = std::make_unique<ShieldTechnoClass>(pTo);
-	pToExt->ShieldData->HP = int(pFromExt->ShieldData->GetShieldRatio() * pToTypeExt->Shield_Strength);
+// Is used for DeploysInto/UndeploysInto
+void ShieldClass::SyncShieldToAnother(TechnoClass* pFrom, TechnoClass* pTo)
+{
+	const auto pFromExt = TechnoExt::ExtMap.Find(pFrom);
+	const auto pToExt = TechnoExt::ExtMap.Find(pTo);
+	if (pFromExt->ShieldData) {
```
Comment (Metadorius) [inline src/Misc/Shield.cpp:71]: Missing newline here and after that code block, brace not on a newline yet again.

File: src/Misc/Shield.cpp [hunk 9]
```diff
@@ -121,312 +120,367 @@ int ShieldTechnoClass::ReceiveDamage(args_ReceiveDamage* args)
...
-	{
-		auto LostHP = this->GetExt()->Shield_Strength - this->HP;
-		if (!LostHP)
+		const int nLostHP = this->Type->Strength - this->HP;
+		if (!nLostHP)
 		{
 			auto result = *args->Damage;
-			if (*args->Damage * GeneralUtils::GetWarheadVersusArmor(args->WH, 
+			if (*args->Damage * GeneralUtils::GetWarheadVersusArmor(args->WH,
 				static_cast<int>(this->Techno->GetTechnoType()->Armor)) > 0)
```
Comment (Metadorius) [inline src/Misc/Shield.cpp:130]: Either don't split a condition or use braces. Right now it's badly readable. Also missing newline before return.

File: src/Misc/Shield.cpp [hunk 10]
```diff
@@ -121,312 +120,367 @@ int ShieldTechnoClass::ReceiveDamage(args_ReceiveDamage* args)
...
+
+		const auto timer = (this->HP <= 0) ? &this->Timers.Respawn : &this->Timers.SelfHealing;
+		timer->Pause();
+
+		this->CloakCheck();
+
+		if (this->IdleAnim)
+		{
+			switch (this->Type->IdleAnim_TemporalAction)
+			{
```
Comment (Metadorius) [inline src/Misc/Shield.cpp:201]: Not sure why the code block content isn't indented.

File: src/Misc/Shield.cpp [hunk 11]
```diff
@@ -121,312 +120,367 @@ int ShieldTechnoClass::ReceiveDamage(args_ReceiveDamage* args)
...
+	if (this->Temporal)
 		return;
 
-	this->DrawShield();
+	this->OnlineCheck();
+
 	this->RespawnShield();
 	this->SelfHealing();
+
+	if ( !this->Cloak && !this->Temporal && this->Online && this->HP > 0 )
```
Comment (Metadorius) [inline src/Misc/Shield.cpp:248]: We don't use spaces around conditions.

File: src/Misc/Shield.cpp [hunk 12]
```diff
@@ -121,312 +120,367 @@ int ShieldTechnoClass::ReceiveDamage(args_ReceiveDamage* args)
...
+	if (!this->Techno || this->Techno->InLimbo || this->Techno->IsImmobilized || this->Techno->Transporter)
+		return;
 
-	if (!this->Update)
+	this->UpdateType();
+
+	this->CloakCheck();
+
+	this->ConvertCheck();
+	if (!this->Available)
```
Comment (Metadorius) [inline src/Misc/Shield.cpp:236]: Missing newline here and excessive use if newlines in all the other function code. I won't mention newlines from this comment on.

File: src/Misc/Shield.cpp [hunk 13]
```diff
@@ -121,312 +120,367 @@ int ShieldTechnoClass::ReceiveDamage(args_ReceiveDamage* args)
...
+		const auto pFrom_TechnoType = TechnoTypeClass::Find(this->TechnoID);
+		const auto pFromType = TechnoTypeExt::ExtMap.Find(pFrom_TechnoType)->Shield;
+
+		if (pFromType->IdleAnim.Get() != pType->IdleAnim.Get())
+			this->KillAnim();
+
+		this->HP = (int)round(
+			(double)this->HP /
+			(double)pFromType->Strength *
+			(double)pType->Strength
```
Comment (Metadorius) [inline src/Misc/Shield.cpp:351]: Not sure at all but maybe there's an operator precedence error here, in any case I'd rather be safe and write multiplication first.

File: src/Misc/Shield.h [hunk 14]
```diff
@@ -7,62 +7,80 @@
...
+	bool Available;
+
+	ShieldTypeClass* Type;
+
+	struct Timers
+	{
+		Timers() :
+			SelfHealing{},
+			Respawn{}
+		{ };
```
Comment (Metadorius) [inline src/Misc/Shield.h:80]: Think the semicolon is not necessary here, also empty code blocks should have spaces in them.

File: src/Utilities/TemplateDef.h [hunk 15]
```diff
@@ -473,6 +487,42 @@ namespace detail {
...
+			{
+				parsed = AttachedAnimFlag::Paused;
+			}
+			else if (_strcmpi(str, "pausedtemporal") == 0)
+			{
+				parsed = AttachedAnimFlag::PausedTemporal;
+			}
+			else if (_strcmpi(str, "none"))
+			{
+				Debug::INIParseFailed(pSection, pKey, parser.value(), "Expected a super weapon affected house");
```
Comment (Metadorius) [inline src/Utilities/TemplateDef.h:517]: CnP error yet again.

**Review by Metadorius — CHANGES_REQUESTED** [review]
The code looks good aside from a bunch of minor issues.

**Comment by Metadorius** [comment]
On develop branch (needs to be checked here as well):
image
image

**Comment by Metadorius** [comment]
Shielded infantry gets infected by parasites, which is kinda unintended default behavior (courtesy of @SukaHati)
[video](https://cdn.discordapp.com/attachments/773636942775582720/856858229950251018/Mental_Omega_2021-06-22_19-20-45.mp4)

**Comment by Metadorius** [comment]
Shielded units with an anim are blocking selection of air units over them (possibly Z sort issue that can be fixed via ini):
[video](https://cdn.discordapp.com/attachments/773636942775582720/856807249099227166/whatisthisbug.mp4)

File: src/Misc/Shield.cpp [hunk 16]
```diff
@@ -11,77 +11,127 @@
...
+void ShieldClass::UpdateType()
 {
-	return TechnoTypeExt::ExtMap.Find(this->Techno->GetTechnoType());
+	this->Type = TechnoTypeExt::ExtMap.Find(this->Techno->GetTechnoType())->Shield;
 }
 
-bool ShieldTechnoClass::Load(PhobosStreamReader& Stm, bool RegisterForChange)
+template <typename T1, typename T2>
+bool ShieldClass::Serialize(T1 pThis, T2& Stm)
 {
```
Comment (Metadorius) [inline src/Misc/Shield.cpp:41]: Why is this method static?

File: src/New/Type/ShieldType.cpp [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+		.Process(this->SelfHealing_Rate)
+		.Process(this->AbsorbOverDamage)
+		.Process(this->BracketDelta)
+		.Process(this->IdleAnim_OfflineAction)
+		.Process(this->IdleAnim_TemporalAction)
+		.Process(this->IdleAnim)
+		.Process(this->BreakAnim)
+		.Process(this->RespawnAnim)
+		.Process(this->HitAnim)
+		.Success()
```
Comment (Metadorius) [inline src/New/Type/ShieldType.cpp:76]: Method returns void, is this really needed?

**Review by Metadorius — APPROVED** [review]

File: src/New/Type/ShieldType.cpp (hunk 17 above)
Comment (Belonit) [reply src/New/Type/ShieldType.cpp:59]: yes. Success is not needed here

File: src/Misc/Shield.cpp (hunk 16 above)
Comment (Belonit) [reply src/Misc/Shield.cpp:41]: I had to resort to such a design, since the templates did not solve the const \ not-const type conflict

**Review by Metadorius — APPROVED** [review]

**Pull request merged**